File: MiniDOMFinanziaria/Finanziaria/Tabelle TEG Max/CRigaTEGMax.vb
Imports minidom.Databases
Imports minidom.Sistema

Partial Public Class Finanziaria


    ''' <summary>
    ''' Definizione di una singola riga dei TEG massimi
    ''' </summary>
    ''' <remarks></remarks>
    Public Class CRigaTEGMax
        Inherits DBObject

        Private m_TabellaID As Integer 'ID della tabella a cui appartiene la riga
        Private m_Tabella As CTabellaTEGMax 'Oggetto tabella a cui appartiene la riga
        Private m_ValoreSoglia As Double  'Limite superiore di validità della riga della tabella
        Private m_Coefficienti() As Double  'Valori soglia del TEG espressi per durata

        Public Sub New()
            Me.m_TabellaID = 0
            Me.m_Tabella = Nothing
            Me.m_ValoreSoglia = 0
            ReDim m_Coefficienti(10)
        End Sub

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Public Property IDTabella As Integer
            Get
                Return GetID(Me.m_Tabella, Me.m_TabellaID)
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.IDTabella
                If oldValue = value Then Exit Property
                Me.m_TabellaID = value
                Me.m_Tabella = Nothing
                Me.DoChanged("IDTabella", value, oldValue)
            End Set
        End Property

        Public Property Tabella As CTabellaTEGMax
            Get
                If (Me.m_Tabella Is Nothing) Then Me.m_Tabella = minidom.Finanziaria.TabelleTEGMax.GetItemById(Me.m_TabellaID)
                Return Me.m_Tabella
            End Get
            Set(value As CTabellaTEGMax)
                Dim oldValue As CTabellaTEGMax = Me.Tabella
                If (oldValue = value) Then Exit Property
                Me.m_Tabella = value
                Me.m_TabellaID = GetID(value)
                Me.DoChanged("Tabella", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Limite superiore di validità della riga della tabella
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property ValoreSoglia As Double
            Get
                Return Me.m_ValoreSoglia
            End Get
            Set(value As Double)
                Dim oldValue As Double = Me.m_ValoreSoglia
                If (oldValue = value) Then Exit Property
                Me.m_ValoreSoglia = value
                Me.DoChanged("ValoreSoglia", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Valori soglia del TEG espressi per durata
        ''' </summary>
        ''' <param name="durata"></param>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Coefficiente(ByVal durata As Integer) As Double
            Get
                Return Me.m_Coefficienti(Fix(durata / 12))
            End Get
            Set(value As Double)
                Dim oldValue As Double = Me.m_Coefficienti(Fix(durata / 12))
                If (oldValue = value) Then Exit Property
                Me.m_Coefficienti(Fix(durata / 12)) = value
                Me.DoChanged("Coefficiente", value, oldValue)
            End Set
        End Property

        Public Overrides Function GetTableName() As String
            Return "tbl_FIN_TEGMaxI"
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("Tabella", Me.IDTabella)
            writer.WriteAttribute("Soglia", Me.m_ValoreSoglia)
            MyBase.XMLSerialize(writer)
            writer.WriteTag("Coefficienti", Me.m_Coefficienti)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "Tabella" : Me.m_TabellaID = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Soglia" : Me.m_ValoreSoglia = XML.Utils.Serializer.DeserializeDouble(fieldValue)
                Case "Coefficienti" : Me.m_Coefficienti = Arrays.Convert(Of Double)(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select

        End Sub

        Protected Overrides Function LoadFromRecordset(ByVal reader As DBReader) As Boolean
            Me.m_TabellaID = reader.Read("Tabella", Me.m_TabellaID)
            Me.m_ValoreSoglia = reader.Read("ValoreSoglia", Me.m_ValoreSoglia)
            For i As Integer = 1 To 10
                Me.m_Coefficienti(i) = reader.Read("Coeff" & (i * 12), Me.m_Coefficienti(i))
            Next
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(ByVal writer As DBWriter) As Boolean
            Dim i As Integer
            writer.Write("Tabella", GetID(Me.m_Tabella, Me.m_TabellaID))
            writer.Write("ValoreSoglia", Me.m_ValoreSoglia)
            For i = 1 To 10
                writer.Write("Coeff" & (i * 12), Me.m_Coefficienti(i))
            Next
            Return MyBase.SaveToRecordset(writer)
        End Function

        Public Overrides Function ToString() As String
            Return Me.m_ValoreSoglia
        End Function

        Public Function Check(ByVal offerta As COffertaCQS) As Boolean
            Throw New NotImplementedException
        End Function

        Protected Overrides Function GetConnection() As CDBConnection
            Return Finanziaria.Database
        End Function


    End Class


End Class

File: MiniDOMFinanziaria/Finanziaria/Tabelle TEG Max/CRigaTEGMax.cs
using System;
using DMD;
using DMD.XML;
using minidom;
using static minidom.Sistema;

namespace minidom
{
    public partial class Finanziaria
    {


        /// <summary>
    /// Definizione di una singola riga dei TEG massimi
    /// </summary>
    /// <remarks></remarks>
        public class CRigaTEGMax : Databases.DBObject
        {
            private int m_TabellaID; // ID della tabella a cui appartiene la riga
            private CTabellaTEGMax m_Tabella; // Oggetto tabella a cui appartiene la riga
            private double m_ValoreSoglia;  // Limite superiore di validità della riga della tabella
            private double[] m_Coefficienti;  // Valori soglia del TEG espressi per durata

            public CRigaTEGMax()
            {
                m_TabellaID = 0;
                m_Tabella = null;
                m_ValoreSoglia = 0d;
                m_Coefficienti = new double[11];
            }

            public override CModulesClass GetModule()
            {
                return null;
            }

            public int IDTabella
            {
                get
                {
                    return DBUtils.GetID(m_Tabella, m_TabellaID);
                }

                set
                {
                    int oldValue = IDTabella;
                    if (oldValue == value)
                        return;
                    m_TabellaID = value;
                    m_Tabella = null;
                    DoChanged("IDTabella", value, oldValue);
                }
            }

            public CTabellaTEGMax Tabella
            {
                get
                {
                    if (m_Tabella is null)
                        m_Tabella = TabelleTEGMax.GetItemById(m_TabellaID);
                    return m_Tabella;
                }

                set
                {
                    var oldValue = Tabella;
                    if (oldValue == value)
                        return;
                    m_Tabella = value;
                    m_TabellaID = DBUtils.GetID(value);
                    DoChanged("Tabella", value, oldValue);
                }
            }

            /// <summary>
        /// Limite superiore di validità della riga della tabella
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
            public double ValoreSoglia
            {
                get
                {
                    return m_ValoreSoglia;
                }

                set
                {
                    double oldValue = m_ValoreSoglia;
                    if (oldValue == value)
                        return;
                    m_ValoreSoglia = value;
                    DoChanged("ValoreSoglia", value, oldValue);
                }
            }

            /// <summary>
        /// Valori soglia del TEG espressi per durata
        /// </summary>
        /// <param name="durata"></param>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
            public double get_Coefficiente(int durata)
            {
                return m_Coefficienti[(int)Maths.Floor(durata / 12d)];
            }

            public void set_Coefficiente(int durata, double value)
            {
                double oldValue = m_Coefficienti[(int)Maths.Floor(durata / 12d)];
                if (oldValue == value)
                    return;
                m_Coefficienti[(int)Maths.Floor(durata / 12d)] = value;
                DoChanged("Coefficiente", value, oldValue);
            }

            public override string GetTableName()
            {
                return "tbl_FIN_TEGMaxI";
            }

            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Tabella", this.IDTabella);
                writer.WriteAttribute("Soglia", this.m_ValoreSoglia);
                base.XMLSerialize(writer);
                writer.WriteTag("Coefficienti", this.m_Coefficienti);
            }

            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Tabella":
                        {
                            this.m_TabellaID = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Soglia":
                        {
                            this.m_ValoreSoglia = (double)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                            break;
                        }

                    case "Coefficienti":
                        {
                            this.m_Coefficienti = (double[])DMD.Arrays.Convert<double>(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            protected override bool LoadFromRecordset(DBReader reader)
            {
                this.m_TabellaID = reader.Read("Tabella", this.m_TabellaID);
                this.m_ValoreSoglia = reader.Read("ValoreSoglia", this.m_ValoreSoglia);
                for (int i = 1; i <= 10; i++)
                    this.m_Coefficienti[i] = reader.Read("Coeff" + i * 12, this.m_Coefficienti[i]);

                return base.LoadFromRecordset(reader);
            }

            protected override bool SaveToRecordset(DBWriter writer)
            {
                int i;
                writer.Write("Tabella", DBUtils.GetID(this.m_Tabella, this.m_TabellaID));
                writer.Write("ValoreSoglia", this.m_ValoreSoglia);
                for (i = 1; i <= 10; i++)
                    writer.Write("Coeff" + i * 12, this.m_Coefficienti[i]);
                return base.SaveToRecordset(writer);
            }

            public override string ToString()
            {
                return m_ValoreSoglia.ToString();
            }

            public bool Check(COffertaCQS offerta)
            {
                throw new NotImplementedException();
            }

            protected override Databases.CDBConnection GetConnection()
            {
                return Database;
            }
        }
    }
}